Expand module nfs3

lordcirth opened the pull request:
### What does this PR do?
Adds add_export() and reload_exports() to modules/nfs3.py

### Tests written?

No

I can add the versionadded tag to docstring if needed.
If this is accepted I intend to write a nfs exports state soon.

lordcirth committed:
nfs3: create add_exports() function

lordcirth committed:
nfs3: add reload_exports() function

This should be all the execution modules needed
to write an nfs exports State.

salt-jenkins commented:
@lordcirth, thanks for your PR! By analyzing the history of the files in this pull request, we identified @techhat, @s0undt3ch and @terminalmage to be potential reviewers.

lordcirth committed:
Merge branch 'develop' into nfs-exports

lordcirth committed:
nfs3: fix bug in list_exports() with similar lines

Before, defining multiple exports with the same path,
which is valid in /etc/exports, would result in
all but the last one being lost.

lordcirth committed:
Merge branch 'develop' into nfs-exports

cachedout commented on salt/modules/nfs3.py:
```diff
@@ -71,6 +75,22 @@ def del_export(exports='/etc/exports', path=None):
     _write_exports(exports, edict)
     return edict
 
+def add_export(exports='/etc/exports', path=None, hosts=None, options=['ro']):
```
We can't have mutable type like lists as default keyword arguments.

lordcirth replied:
So, should "options" not have a default value?

lordcirth committed:
Merge branch 'develop' into nfs-exports

lordcirth committed:
nfs3.add_export(): fix use of list as default argument

lordcirth commented:
Got a pre-existing bug to fix in add_exports().  It allows specifying multiple hosts and writes them comma-separated to the file.  This is not supported by /etc/exports.  A solution might be to copy the options list into a new export for each host.

lordcirth committed:
nfs3: bugfix: /etc/exports does not support comma-sep hosts

lordcirth commented:
Fixed the bug by just changing the host list to string everywhere.  Left typechecks to catch anything that didn't get changed.

lordcirth committed:
nfs3: allow multiple exports of the same path

cachedout commented:
Hi @lordcirth. Thanks for those recent changes. There are a few lint errors that also need to be fixed, please:

https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/12035/violations/file/salt/modules/nfs3.py/

lordcirth committed:
nfs3: linting

lordcirth commented:
It seems like there is a unit test for list_exports() which has failed since hosts: is no longer a list.  Should we just change the test?

cachedout commented:
Go Go Jenkins!

cachedout commented:
@lordcirth Yes, if the output of the command has changed then the test needs to be modified to reflect that. Thanks.

garethgreenaway committed:
Merge branch 'develop' into nfs-exports

lordcirth committed:
Merge branch 'develop' into nfs-exports

lordcirth committed:
test_nfs3: pass string, not list

Commit:
Merge branch 'develop' into nfs-exports

rallytime commented:
Hi @lordcirth - Can you clean up these final lint errors? 

https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/13734/violations/file/salt/modules/nfs3.py/

Then we can get this in. :)

lordcirth committed:
nfs3: fix 3 lint errors

lordcirth committed:
Merge branch 'nfs-exports' of https://github.com/lordcirth/salt into nfs-exports

lordcirth commented:
I'd just like to mention that this list of CI runs doesn't present much easy-to-find information to contributors.  While I don't know much about Jenkins, it would be nice if the CI could be made more accessible so that contributors could see lint errors and such without wasting maintainer's time digging them up.

cachedout commented:
@lordcirth Yes, I agree. We have some projects in the works to improve in that area. :] Thanks for the reminder, however. It really does need improvement.

Commit:
Merge branch 'develop' into nfs-exports
